dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/anr/ANRDetectorRunnable.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.anr
+
+import android.os.Handler
+import com.datadog.android.rum.GlobalRum
+import com.datadog.android.rum.RumErrorSource
+
+/**
+ * A Runnable running on a background thread detecting ANR on the main thread.
+ *
+ * It runs in a background thread and schedules regular no-op
+ */
+internal class ANRDetectorRunnable(
+    private val handler: Handler,
+    private val anrThresholdMs: Long = ANR_THRESHOLD_MS,
+    private val anrTestDelayMs: Long = ANR_TEST_DELAY_MS
+) : Runnable {
+
+    private var shouldStop = false
+
+    // region Runnable
+
+    override fun run() {
+
+        while (!Thread.interrupted()) {
+            if (shouldStop) return
+
+            try {
+                val callback = CallbackRunnable()
+
+                synchronized(callback) {
+                    handler.post(callback)
+                    callback.wait(anrThresholdMs)
+
+                    if (!callback.wasCalled()) {
+                        GlobalRum.get().addError(
+                            ANR_MESSAGE,
+                            RumErrorSource.SOURCE,
+                            ANRException(handler.looper.thread),
+                            emptyMap()
+                        )
+                        callback.wait()
+                    }
+                }
+
+                Thread.sleep(anrTestDelayMs)
+            } catch (e: InterruptedException) {
+                break
+            }
+        }
+    }
+
+    // endregion
+
+    @Synchronized
+    fun stop() {
+        shouldStop = true
+    }
+
+    // We need to let this class extend java's Object
+    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
+    internal class CallbackRunnable : Object(), Runnable {
+
+        private var called = false
+
+        @Synchronized
```

Comment:
We need the `@Synchronized` here and below to ensure the read/write operation on `called` are synchronized between threads.
This run operation will never block for too long (it's just an assignment of a boolean and a notify all).

Reply:
I see, but `called` is a boolean primitive which is `volatile` by default. IMO there is no need for synchronization here as this is already handled by JVM, the problem comes only when you want to increment a number or modify a boolean only if the previous state has a specific value. Exceptionally you could use an AtomicBoolean here.

Reply:
Actually I need to make this `@Synchronized` otherwise the `notifyAll` will fail

Reply:
I'm not 100% sure how does it work. `@Synchronized` here is effectively `synchronized(this)` (because synchronized methods are using instance as a lock, unless Kotlin is doing something different under the hood and not using the instance for synchronization), which is the same as `synchronized(callback)` above (same instance), just called from different thread. But afaik one thread (thread of `CallbackRunnable`) cannot enter the monitor which is held by another thread (thread of `ANRDetectorRunnable`). So in this case how handler thread is able to enter `run` method? probably I miss something.